### PR TITLE
[Google Blockly] Add firehose logging for xml/json block comparison

### PR DESCRIPTION
**Pull request by mikeharv**
## Context
This is part of the work to move towards using JSON to save/load student Blockly code in our labs. For more information, see the [dev doc](https://docs.google.com/document/d/1V367MAAYU0JXETVX6l48_6HW2h93WDzdljJc1ohLZPc/edit#).

This PR is to help us confirm that Blockly's JSON serialization of a workspace is equivalent to the XML encoding we currently store. This is the first step of the efforts to move to JSON serialization as the default.

## Proposal
In order to build confidence that Blockly's JSON serialization of one of our workspace is equivalent to the current XML encoding,  the new `testJsonSerialization()` function can do the following for a given workspace:
- call `Blockly.Xml.blockSpaceToDom` to get the XML elements we would normally save
- call `Blockly.serialization.workspaces.save` to get the JON serialization we are interested in
- for each structure above, load the blocks into a temporary workspace that is never rendered (using `Blockly.Xml.domToWorkspace` and `Blockly.serialization.workspaces.load`, respectively.
- for each temporary workspace, get an array of the blocks that were created.
- compare each block array and test whether the actual block objects are deeply equal or not.
- log the results to Firehose/Redshift (for now) or CloudWatch (once that work is ready)

Most of the steps above are quite straight-forward. However, comparing the block arrays is complicated because each contains objects that are deeply nested and include references to other blocks in the array, references to the parent block itself (circular references), and references to the temporary workspaces on which they were created (which we don't actually care about). This branch includes `compareBlockArrays()` which is designed check that the two arrays are deeply equal while accounting for the references that would potentially create infinite loops or lead to false negatives.

The function returns an array of "differences" that can be logged to Firehose, if there are any. 

A difference object in this context contains information about a difference found between the two block objects being compared. It has the following properties:

- `result`: a string indicating the type of difference found. It can be one of the following: 'different length', 'different keys', 'different types', 'different values', or 'non-object'.
- `path`: a string indicating the path to the property that has a difference. The path is a dot-separated string of keys that lead to the property.
- Depending on the type of difference, a difference object will contain additional properties, either keys, values or type descriptions that explain the rationale.

We can collect data at a sample rate, such as 5%, randomly calculated whenever a user loads a project or presses Run. An experiment flag can also be used to force Firehose to log the record each time, as well as to load it to the console to aid in internal debugging.

## Sample Records
Each record will include a link to the project URL which will allow us to access and manual test for any problematic blocks. It will also include a reference to the current Blockly version which should help us to be aware if a future version potentially introduces a regression to this logic.

In the event that no differences are found between the block arrays, nothing is logged.

In the event that differences are found, the record will include these differences in an array. Somewhat conveniently, the recent 9.3 upgrade included a fix that will ensure blocks' `movable`/`deletable` states are serialized as we'd expect. By downgrading to 9.1 locally, we can see that the new function finds differences that we'd potentially need to address.

```
{
    "projectUrl": "http://localhost-studio.code.org:3000/projects/dance/qgYVrbNfZyFEjQWmpkRwgQ",
    "blocklyVersion": "9.1.1",
    "differences": [
        {
            "result": "different values",
            "path": ".0",
            "key": "deletable_",
            "value1": false,
            "value2": true
        },
        {
            "result": "different values",
            "path": ".0",
            "key": "movable_",
            "value1": false,
            "value2": true
        }
    ]
}
```

Above, we see that a different value was found for the keys `deletable_` and `movable_` at the path ".0", referring to the first block in the array. Our first block is usually a `when_run` block that is set to be neither movable nor deletable and the new comparison function reveals that the JSON serializer missed these values when running Blockly 9.1.1. 
Note that paths will be longer for more deeply nested values.

## Redshift Queries

Records can be queried at https://us-east-1.console.aws.amazon.com/sqlworkbench/home?region=us-east-1#/client/

A starting query might look like this:
```
SELECT
    *
FROM
    "dashboard"."analysis"."events"
WHERE environment = 'production'
and study = 'blockly-json'
LIMIT 100;
```
 Here are some examples of how this might look. Note that the screenshots are of query for the `'development'` environment.
image
image
The table can be exported as a JSON/CSV file. It's also easy to filter directly in Redshift for things like the lab type (e.g. "dance", "flappy"), or a particular Blockly version number.

## Links

https://codedotorg.atlassian.net/browse/SL-640

## Deployment strategy

Dependent upon how much good data we end up collecting, we can adjust the sample rate or turn the entire feature off easily with a DCDO flag. We are not concerned with the cost of recording samples, but it's possible that there could be a bottleneck in place that could result in dropping samples if Firehose is past its capacity. For this project, we likely would not move forward until the test is succeeding for all projects, and we can assume that a using a smaller sample rate will still be adequate to make that determination.

## Privacy

We will collect project links that will have been recently opened by our users.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/blockly/addons/cdoUtils.js [hunk 1]
```diff
@@ -106,9 +109,168 @@ export function getUserTheme(themeOption) {
...
+    testJsonSerialization(workspace);
+  }
   return Blockly.Xml.domToText(Blockly.Xml.blockSpaceToDom(workspace));
-  // After supporting JSON block sources, change to:
-  // return JSON.stringify(Blockly.serialization.workspaces.save(workspace));
+}
+
+function testJsonSerialization(workspace) {
+  const FIREHOSE_STUDY = 'blockly-json';
+  const FIREHOSE_EVENT_PASS = 'blocks-identical';
```
Comment (ebeastlake) [inline apps/src/blockly/addons/cdoUtils.js]: nit/question for the group: Is there any advantage to having these names mirror each other, something like...
```
const FIREHOSE_STUDY = 'blockly-blocks-omparison';
const FIREHOSE_EVENT_PASS = 'blockly-blocks-identical';
const FIREHOSE_EVENT_FAIL = 'blockly-blocks-different';
```
Or something? I don't have context on this PR, but I wonder if that could lead to some tiny advantage like being able to query using syntax like 
`WHERE event LIKE 'blockly-blocks-%'`. (And maybe this only applies to the pass/failure cases and the study name doesn't matter.)

File: apps/src/blockly/addons/cdoUtils.js [hunk 2]
```diff
@@ -106,9 +109,168 @@ export function getUserTheme(themeOption) {
 }
 
 export function getCode(workspace) {
+  // Begin collecting data about our ability to correctly serialize a workspace in JSON format.
+  if (
+    experiments.isEnabled(experiments.BLOCKLY_JSON) ||
+    Math.random() <= 0.05
```
Comment (ebeastlake) [inline apps/src/blockly/addons/cdoUtils.js]: Is there an experiment config we could use to modify this value without a deploy (ex. `DCDO.blocky_experiment_sample_rate`), just in case we want to be able to modify without a code change? (Is that worth doing?) Otherwise, could we make it a constant?

File: apps/src/blockly/addons/cdoUtils.js [hunk 3]
```diff
@@ -106,9 +109,168 @@ export function getUserTheme(themeOption) {
...
+      study: FIREHOSE_STUDY,
+      event: result,
+      data_json: JSON.stringify(recordData),
+    },
+    {
+      alwaysPut: true, // Allows logging from development environments
+      includeUserId: true,
+    }
+  );
+  if (experiments.isEnabled(experiments.BLOCKLY_JSON)) {
```
Comment (ebeastlake) [inline apps/src/blockly/addons/cdoUtils.js:168]: Can you clarify here: this experiment is going to be released for everyone, right (at least 5% of the time)? But your PR description says, "An experiment flag can also be used... to load it to the console to aid in internal debugging." Does that mean the `experiments.BLOCKLY_JSON` refers to an extra, optional experimental flag on top of the experiment that will be sampling in production?

File: apps/src/blockly/addons/cdoUtils.js [hunk 4]
```diff
@@ -106,9 +109,168 @@ export function getUserTheme(themeOption) {
...
+  // references.
+  const visited = new Set();
+  // Some block properties are expected to hold different values, or refer to blocks already in the array.
+  const keysToSkip = [
+    'parentBlock_', // a circular reference to another block in the array
+    'sourceBlock_', // a circular reference to another block in the array
+    'workspace', // a reference to the unique temporary workspaces used to create the two block arrays
+  ];
+
+  function compareValues(arg1, arg2, path = '', differences = []) {
```
Comment (ebeastlake) [inline apps/src/blockly/addons/cdoUtils.js:189]: Can/should we include a test suite for `compareValues` and `compareBlockArrays` with this PR? It seems like there are a lot of well-thought-out edge cases here.

File: apps/src/blockly/addons/cdoUtils.js [hunk 5]
```diff
@@ -106,9 +109,168 @@ export function getUserTheme(themeOption) {
...
+        differences.push({
+          result: 'different length',
+          path: path,
+          key1: keys1.length,
+          key2: keys2.length,
+        });
+      }
+
+      for (let i = 0; i < keys1.length; i++) {
+        const key1 = keys1[i];
```
Comment (fisher-alice) [inline apps/src/blockly/addons/cdoUtils.js:215]: In the array returned from the call to `getAllBlocks`, are the blocks always returned in the same sequence? I'm wondering about detached/unused blocks or programs with 2 or more event blocks.

File: apps/src/blockly/addons/cdoUtils.js [hunk 6]
```diff
@@ -1,8 +1,11 @@
 import {ToolboxType, CLAMPED_NUMBER_REGEX, DEFAULT_SOUND} from '../constants';
 import cdoTheme from '../themes/cdoTheme';
+import experiments from '@cdo/apps/util/experiments';
+import firehoseClient from '@cdo/apps/lib/util/firehose';
 import {APP_HEIGHT} from '@cdo/apps/p5lab/constants';
 import {SOUND_PREFIX} from '@cdo/apps/assetManagement/assetPrefix';
 
+
```
Comment (fisher-alice) [inline apps/src/blockly/addons/cdoUtils.js]: nit: unnecessary extra line

File: apps/src/blockly/addons/cdoUtils.js (hunk 4 above)
Comment (fisher-alice) [reply apps/src/blockly/addons/cdoUtils.js:189]: +1 for adding unit tests

**Review by fisher-alice — COMMENTED** [review]
This is a significant step towards using json to saving student sources!
I pulled down your branch and tried to use lodash's `isEqual` which did NOT work so nice job customizing the comparison of the block arrays!

I had a question about the way block arrays are saved and +1 to @ebeastlake suggestion to include tests. This is exciting!

File: apps/src/blockly/addons/cdoUtils.js (hunk 4 above)
Comment (bencodeorg) [reply apps/src/blockly/addons/cdoUtils.js:189]: Is there no existing solution out there that will compare objects (and handle circular references)?

Comment (mikeharv) [reply apps/src/blockly/addons/cdoUtils.js:189]: There very well could be. We also need to exclude certain properties from the comparison (e.g. `workspace`), and we want to be able to log the specific differences that are found so that we can effectively debug. For example, if the checker had an early return at the first difference found, we would potentially miss that a whole bunch of differences could be present for a given workspace. For these reasons, it seemed best just to implement my own utility. Also, I'm planning to add unit tests in the morning that should help prove that this utility does what it's meant to. The test will include cases for the exceptions we care about.

File: apps/src/blockly/addons/cdoUtils.js [hunk 7]
```diff
@@ -106,9 +108,167 @@ export function getUserTheme(themeOption) {
 }
 
 export function getCode(workspace) {
+  // Begin collecting data about our ability to correctly serialize a workspace in JSON format.
+  if (
+    experiments.isEnabled(experiments.BLOCKLY_JSON) ||
+    Math.random() <= 0.05
+  ) {
+    testJsonSerialization(workspace);
```
Comment (bencodeorg) [inline apps/src/blockly/addons/cdoUtils.js]: Can we do this comparison asynchronously (sorry if I'm missing discussion of this somewhere)?

Comment (mikeharv) [reply apps/src/blockly/addons/cdoUtils.js]: Not really, unless we find a way to also spin up an instance of the Blockly lab with the appropriate custom blocks installed. Realistically, we'd need the actual curriculum level hosts the project to be sure that we're able to create all of the blocks.

File: apps/src/blockly/addons/cdoUtils.js (hunk 3 above)
Comment (mikeharv) [reply apps/src/blockly/addons/cdoUtils.js:168]: The new utility will run for all users, 5% of the time. If we want to do additional testing ourselves, we can use the experiment flag which will force the utility to run 100% of the time until disabled. Additionally, I've added a DCDO flag as a first check. Unless `DCDO.get('blockly-json-experiment', true)`, we generate or compare blocks for anyone. This should give us some added comfort in case anything unexpected happens (performance impacts to users or overwhelming Firehose).

File: apps/src/blockly/addons/cdoUtils.js (hunk 4 above)
Comment (mikeharv) [reply apps/src/blockly/addons/cdoUtils.js:189]: Unit tests have been added.

File: apps/src/blockly/addons/cdoUtils.js (hunk 5 above)
Comment (mikeharv) [reply apps/src/blockly/addons/cdoUtils.js:215]: They should be, and I've tested with workspaces with multiple top-level blocks.  If they were not in the same order for any reason, the utility will return a difference.

File: apps/src/blockly/addons/cdoUtils.js (hunk 7 above)
Comment (bencodeorg) [reply apps/src/blockly/addons/cdoUtils.js]: I just mean can we do this in the background (and continue to load the page)? As written, this will block actually loading the workspace, yeah?

File: apps/src/blockly/addons/cdoUtils.js [hunk 8]
```diff
@@ -106,9 +109,167 @@ export function getUserTheme(themeOption) {
...
+  // references.
+  const visited = new Set();
+  // Some block properties are expected to hold different values, or refer to blocks already in the array.
+  const keysToSkip = [
+    'parentBlock_', // a circular reference to another block in the array
+    'sourceBlock_', // a circular reference to another block in the array
+    'workspace', // a reference to the unique temporary workspaces used to create the two block arrays
+  ];
+
+  function compareValues(arg1, arg2, path = '', differences = []) {
```
Comment (bencodeorg) [inline apps/src/blockly/addons/cdoUtils.js:189]: Does this become a generic "compare objects" function if we pass in `keysToSkip` as an argument? Might be nice to isolate this as a utility.

Comment (fisher-alice) [reply apps/src/blockly/addons/cdoUtils.js:189]: Cool idea @bencodeorg

File: apps/test/unit/blockly/cdoUtilsTest.js [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+
+    it('should skip keysToSkip when comparing', () => {
+      const differences = compareBlockArrays(
+        xmlBlocksWithParent,
+        jsonBlocksWithParent
+      );
+      expect(differences).to.deep.equal([]);
+    });
+  });
+});
```
Comment (fisher-alice) [inline apps/test/unit/blockly/cdoUtilsTest.js:187]: Could add another unit test to check if it skips more than one key for keys that are in `keysToSkip`, e.g.  'parentBlock_' and     'sourceBlock_'.

File: apps/src/blockly/addons/cdoUtils.js (hunk 4 above)
Comment (jmkulwik) [reply apps/src/blockly/addons/cdoUtils.js:189]: One thought: Do we need unit tests? (sorry, I know I'm late to the game here)

I usually think of unit tests as something that keeps us from causing regressions when we change code in the future. But if I'm reading this right, this util is intended to be used only for a short period of time. We'll use it to run some tests, but then it will be deleted. It's similar to a one-off script. Because of that, we won't have future maintainers, and we won't have a risk of future regression on this code.

**Comment by jmkulwik** [comment]
(Note this is intended as a time-saving suggestion. Feel free to ignore if it'll slow things down.)
Do you think that testing across a few thousand student projects will give us a significantly higher confidence than testing across say 10 or 20? Would this work instead:
- Grab 10 decently complicated projects from the project gallery
- Duplicate them locally
- Load them on the local dev environment with this script running
- (If it makes sense to do so at the same time) visually spot check the XML version and the JSON version

Doing this with 1 or 2 of Brad's dance party projects, for example, would make me feel very confident that the JSON loading works as expected in Dance. Perhaps the same could be said for other labs that have been migrated too.

**Comment by mikeharv** [comment]
> (Note this is intended as a time-saving suggestion. Feel free to ignore if it'll slow things down.) Do you think that testing across a few thousand student projects will give us a significantly higher confidence than testing across say 10 or 20? Would this work instead:
> 
> * Grab 10 decently complicated projects from the project gallery
> * Duplicate them locally
> * Load them on the local dev environment with this script running
> * (If it makes sense to do so at the same time) visually spot check the XML version and the JSON version
> 
> Doing this with 1 or 2 of Brad's dance party projects, for example, would make me feel very confident that the JSON loading works as expected in Dance. Perhaps the same could be said for other labs that have been migrated too.

@jmkulwik I've actually already done this, including with one of Brad’s dance projects with 106 blocks.  I’ve yet to find any differences in any project I copy over.

File: apps/src/blockly/addons/cdoUtils.js (hunk 4 above)
Comment (breville) [reply apps/src/blockly/addons/cdoUtils.js:189]: > I usually think of unit tests as something that keeps us from causing regressions when we change code in the future.

I think there's another category of unit tests which are built at the same time (or even before) the code (classically in [TDD](https://en.wikipedia.org/wiki/Test-driven_development)), which are used to help in the development of code and increase confidence that it works, even if it's only temporary.  So for that reason, it seems good to have some here.

**Comment by breville** [comment]
> I’ve yet to find any differences in any project I copy over.

Hopefully we can find (or simulate) a difference at some point to increase confidence that the error reporting path works in production? :)

**Comment by mikeharv** [comment]
> > I’ve yet to find any differences in any project I copy over.
> 
> Hopefully we can find (or simulate) a difference at some point to increase confidence that the error reporting path works in production? :)

I’ve done that as well. I’ve simulated differences by feeding it mismatched data directly, similar to the unit tests. Additionally, prior to 9.3, mainline Blockly deliberately wasn’t serializing a blocks movable/deleteable states unless you you used an additional plug-in. By locally downgrading our Blockly version you’ll get consistent difference on blocks like “when run” that are usually marked as immovable and undeleteable in the curriculum. The PR description includes screens of output for this, which should build confidence that it would find other errors, if present.

File: apps/src/blockly/addons/cdoUtils.js (hunk 7 above)
Comment (mikeharv) [reply apps/src/blockly/addons/cdoUtils.js]: Sorry for misunderstanding. The test is now being run asynchronously. Great suggestion!

File: apps/src/blockly/addons/cdoUtils.js (hunk 1 above)
Comment (mikeharv) [reply apps/src/blockly/addons/cdoUtils.js]: Since we’re not going to log successful tests, we no longer need two different events.

File: apps/src/blockly/addons/cdoUtils.js (hunk 2 above)
Comment (mikeharv) [reply apps/src/blockly/addons/cdoUtils.js]: I’ve added a DCDO Boolean flag that can be used to turn the entire feature on or off. I went with that because it seems to be how we usually use DCDO flags and that made me feel it would be most intuitive for any engineer that suddenly needed to go in and change the value.  I could also modify this to be a rate like you suggest if that seems better, so let me know what you think.

File: apps/src/blockly/addons/cdoUtils.js (hunk 3 above)
Comment (mikeharv) [reply apps/src/blockly/addons/cdoUtils.js:168]: I added a second DCDO value for the sample rate (0.0 - 1.0). In order for any logging to begin, we would need to set `'blockly-json-experiment'` to true and also set `'blockly-json-sample-rate'` to a value like 0.05. This should allow us to start with a very low rate, like 1% and gradually increase it if things seem okay. We'll also be able to turn the whole thing off if desired.

File: apps/src/blockly/addons/cdoUtils.js (hunk 4 above)
Comment (mikeharv) [reply apps/src/blockly/addons/cdoUtils.js:189]: I'm inclined to agree with @breville in this. Given that the new utility is so complex, adding unit tests are helpful right now as they give confidence that it does what it is designed to do. Planning to leave the tests in as of right now.

**Review by jmkulwik — APPROVED** [review]
Nice work Mike!

I'm good with this going out. There's been a lot of discussion, so here's my reasoning:

A lot of it revolves around risk vs consequence, where risk is the likelihood of something going wrong, and consequence is the impact of something going wrong.

The risk of the eventual XML -> JSON conversion causing an issue seems to be very low. There have been some hand tests already done against complex projects that have all gone well. Additionally, the nature of the helper scripts provided by Google Blockly also adds to the low-risk profile.

However, the consequence of an issue in the eventual XML -> JSON transition is very high (potential loss of data, etc).

This PR will give us the confidence that the risk is indeed very low. Even running it for a few hours will give us a good measure of the risk. (One recommendation, let's find a way to add an [Elephant in Cairo](https://en.wikipedia.org/wiki/Elephant_in_Cairo) so we are confident that this will report errors as expected).

Thinking about minimizing the consequence, the [JSON Serialization for saving user Blockly workspace state](https://docs.google.com/document/d/1V367MAAYU0JXETVX6l48_6HW2h93WDzdljJc1ohLZPc/edit#) covers the work to minimize consequences. That will involve follow-up work to be done after this PR.

File: apps/test/unit/blockly/cdoUtilsTest.js (hunk 9 above)
Comment (mikeharv) [reply apps/test/unit/blockly/cdoUtilsTest.js:187]: Probably could, but for the purpose of ensuring that this works as expected, it doesn't seem strictly necessary.

File: apps/src/blockly/addons/cdoUtils.js (hunk 8 above)
Comment (mikeharv) [reply apps/src/blockly/addons/cdoUtils.js:189]: Going to call this out of scope for now. :)
